src/tests/Microsoft.PowerFx.Interpreter.Tests/DeferredTypeTests.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.PowerFx.Core.Localization;
+using Microsoft.PowerFx.Core.Tests.Helpers;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Types;
+using Xunit;
+
+namespace Microsoft.PowerFx.Interpreter
+{
+    public class DeferredTypeTests
+    {
+        [Theory]
```

Comment:
>Theory [](http://example.com/codeflow?start=9&length=6)

More tests:
 
Sum(X,X) // both args are deferred 
X.Field1.Field2    // cascading field deref

ForAll([1,2,3], X)
ForAll(ParseJson("[1]"), X) 


If(true, X) // with IF
If(true, X, 1)
If(true, X, X)

Reply:
Added tests except for ForAll(). Since in current iteration, we are not supporting the functions which can take expression as an input.

Reply:
Made changes to allow functions with lambda arg, Added tests for ForAll() as well now.